File: python/pssparser/parser.py
from io import StringIO
from typing import Dict, List, Tuple, TextIO

class ParseException(Exception):
    def __init__(self, message, markers=None):
        super().__init__(message)
        self.markers = markers or []

class Parser(object):

    def __init__(
            self,
            collect_docstrings : bool = False,
            collect_comments : bool = False):
        """Build a parser.

        :param collect_docstrings: attach the doc comment above a declaration
            to it, as ``ScopeChild.docstring``.
        :param collect_comments: capture *every* comment, on statements as
            well as declarations, as ``ScopeChild.comments``. Implies
            ``collect_docstrings``.

        Both default off: collection costs a token-stream walk per construct,
        and most callers only want the tree.
        """
        import pssparser.core as zspp
        import pssparser.ast as zsp_ast
        self.ast_f = zsp_ast.Factory.inst()
        self.parser_f = zspp.Factory.inst()
        self._root = None
        self._filenames : Dict[int,str] = {}
        self._files : List[zsp_ast.GlobalScope] = []
        self._enable_profiling = False
        self._last_builder = None
        self._markers = []
        #: fileid -> path, snapshotted by link() before it clears its state.
        self._file_map : Dict[int,str] = {}
        #: One builder per Parser, created lazily by _mkBuilder.
        self._builder = None
        #: Applied by _mkBuilder to every builder it creates. link() drops the
        #: builder, so these cannot be set once at construction time.
        self._collect_docstrings = collect_docstrings
        self._collect_comments = collect_comments

    def _mkBuilder(self, marker_l):
        """Return this Parser's builder, pointed at *marker_l*.

        The builder is created once and reused, because it accumulates the
        source units it has processed: compile-time expressions are evaluated
        during AST construction and may reference types and constants from a
        previously-processed unit (PSS 3.1 19.1.2).  A builder per parse() call
        would restart that environment, so a second call could not see the
        constants declared by the first.  Each call still gets a fresh marker
        collector, since markers are reported per call.
        """
        if self._builder is None:
            self._builder = self.parser_f.mkAstBuilder(marker_l)
            if self._collect_docstrings:
                self._builder.setCollectDocStrings(True)
            if self._collect_comments:
                self._builder.setCollectComments(True)
        else:
            self._builder.setMarkerListener(marker_l)
        return self._builder

    def parse(self, files : List[str]) -> bool:
        import pssparser.core as zspp
        marker_l = self.parser_f.mkMarkerCollector()
        builder = self._mkBuilder(marker_l)

        if self._enable_profiling:
            builder.setEnableProfile(True)

        file_id = 0
        if len(self._files) == 0:
            stdlib = self.ast_f.mkGlobalScope(len(self._files))
            self.parser_f.loadStandardLibrary(builder, stdlib)
            self._files.append(stdlib)

        for f in files:
            id = len(self._files)
            self._filenames[id] = f
            with open(f, "r") as fp:
                ast = self.ast_f.mkGlobalScope(id)
                builder.build(ast, fp)
            
            if marker_l.hasSeverity(zspp.MarkerSeverityE.Error):
                self._markers = self._collectMarkers(marker_l)
                err = self._mkErrorMessage(marker_l)
                raise ParseException(err, self._markers)

            self._files.append(ast)

        self._last_builder = builder
        self._markers = self._collectMarkers(marker_l)

        return True

    def parses(self, files : List[Tuple[str, str]]) -> bool:
        import pssparser.core as zspp
        marker_l = self.parser_f.mkMarkerCollector()
        builder = self._mkBuilder(marker_l)

        if self._enable_profiling:
            builder.setEnableProfile(True)

        if len(self._files) == 0:
            stdlib = self.ast_f.mkGlobalScope(len(self._files))
            self.parser_f.loadStandardLibrary(builder, stdlib)
            self._files.append(stdlib)

        for fname,fstr in files:
            id = len(self._files)
            self._filenames[id] = fname
            ast = self.ast_f.mkGlobalScope(id)
            builder.build(ast, StringIO(fstr))
            
            if marker_l.hasSeverity(zspp.MarkerSeverityE.Error):
                self._markers = self._collectMarkers(marker_l)
                err = self._mkErrorMessage(marker_l)
                raise ParseException(err, self._markers)

            self._files.append(ast)

        self._last_builder = builder
        self._markers = self._collectMarkers(marker_l)

        return True
    
    def enable_profiling(self, enable: bool = True):
        """Enable or disable ANTLR profiling for subsequent parse operations."""
        # Profiling must be enabled before creating builder
        # This is a placeholder - will take effect on next parse
        self._enable_profiling = enable

    def get_profile_info(self):
        """Get profiling information from the last parse operation.
        
        Returns ParseProfileInfo object with decision-level and aggregate metrics,
        or None if profiling was not enabled or no parse has been performed.
        """
        if hasattr(self, '_last_builder') and self._last_builder is not None:
            return self._last_builder.getProfileInfo()
        return None
    
    def link(self) -> 'zsp_ast.RootSymbolScope':
        import pssparser.core as zspp
        linker = self.parser_f.mkAstLinker()
        marker_l = self.parser_f.mkMarkerCollector()

        ret = linker.link(marker_l, self._files)

        # Collect unconditionally. This ran only inside the `hasSeverity`
        # branch below, so a link that produced warnings but no errors built
        # them, handed them to the collector, and then dropped them: the CLI
        # reads `parser.markers`, which stayed empty, and printed "0 errors in
        # 0 files". Both parse paths above already collect on success; only
        # this one did not.
        self._markers.extend(self._collectMarkers(marker_l))

        if marker_l.hasSeverity(zspp.MarkerSeverityE.Error):
            err = self._mkErrorMessage(marker_l)
            raise ParseException(err, self._markers)

        # Snapshot the fileid -> path mapping before clearing. Consumers that
        # run *after* linking -- checker plug-ins in particular -- still need
        # to resolve a fileid to a path, and clearing without a snapshot is
        # why CheckContext.file_map arrived empty.
        #
        # Only the mapping is retained, never the GlobalScope objects. The
        # linker takes ownership of them (TaskBuildSymbolTree pushes each into
        # root->getUnits() as an owning pointer), so a Python wrapper kept
        # past this point is a second owner of the same memory and will fault.
        # That is why _files is cleared here; it is deliberate, not an
        # oversight. Reach the per-file scopes through the linked root
        # instead -- see user_units().
        self._file_map = dict(self._filenames)
        self._root = ret

        self._filenames.clear()
        self._files.clear()

        # Drop the builder along with the units it was handed.  It holds
        # borrowed pointers to these GlobalScopes as its compile-time
        # environment (PSS 3.1 19.1.2), and ownership has just moved to the
        # linked root: a later link() replaces that root, frees the scopes, and
        # a parse() after that would read freed memory.  A fresh builder
        # restarts the environment, which matches what just happened -- these
        # units are no longer the Parser's to offer.
        self._builder = None

        return ret

    @property
    def file_map(self) -> Dict[int, str]:
        """Map fileid -> source path for the user-supplied files.

        Populated by :meth:`link` and valid afterwards. Built-in and library
        units are excluded: they carry fileid 0 or -1 and have no user path.
        """
        return dict(self._file_map)

    def user_units(self) -> List['zsp_ast.GlobalScope']:
        """The GlobalScope of each user-supplied file, in parse order.

        Read from the linked root, which owns the units after :meth:`link`.
        Built-in and standard-library units are filtered out by fileid.

        Returns an empty list before :meth:`link` has run.
        """
        if self._root is None:
            return []
        out = []
        for i in range(self._root.numUnits()):
            u = self._root.getUnit(i)
            if u is not None and u.getFileid() in self._file_map:
                out.append(u)
        return out


    def _mkErrorMessage(self, marker_l) -> str:
        import pssparser.core as zspp
        prefix = {
            zspp.MarkerSeverityE.Error : "Error: ",
            zspp.MarkerSeverityE.Warn : "Warning: ",
            zspp.MarkerSeverityE.Info : "Info: ",
            zspp.MarkerSeverityE.Hint : "Hint: ",
        }
        msg = ""
        for i in range(marker_l.numMarkers()):
            marker = marker_l.getMarker(i)
            loc = marker.loc()
            marker_m = "%s%s %s:%d:%d" % (
                prefix[int(marker.severity())],
                marker.msg(),
                self._filenames.get(loc.file, "<unknown>"),
                loc.line,
                loc.pos+1)
            msg += marker_m + "\n"

        return msg

    def _collectMarkers(self, marker_l) -> list:
        """Collect structured marker data from a MarkerCollector."""
        import pssparser.core as zspp
        severity_names = {
            int(zspp.MarkerSeverityE.Error): "error",
            int(zspp.MarkerSeverityE.Warn): "warning",
            int(zspp.MarkerSeverityE.Info): "info",
            int(zspp.MarkerSeverityE.Hint): "hint",
        }
        result = []
        for i in range(marker_l.numMarkers()):
            m = marker_l.getMarker(i)
            loc = m.loc()
            filename = self._filenames.get(loc.file, "<unknown>")
            result.append({
                "severity": severity_names.get(int(m.severity()), "unknown"),
                "message": m.msg(),
                "file": filename,
                "line": loc.line,
                "col": loc.pos + 1,
            })
        return result

    @property
    def markers(self) -> list:
        """Structured list of markers from the last parse/link operation.
        
        Each marker is a dict with keys:
            severity: "error", "warning", "info", or "hint"
            message: Human-readable error message
            file: Source filename
            line: Line number (1-based)
            col: Column number (1-based)
        """
        return list(self._markers)

    @property
    def root(self) -> 'zsp_ast.RootSymbolScope':
        return self._root

File: ts/src/Parser.ts
/**
 * The PSS parser.
 *
 * A transliteration of `python/pssparser/parser.py`, not a new design. That
 * file is 380 lines of accumulated judgement about what a pssparser consumer
 * needs -- stdlib auto-loading, builder reuse across parse calls, marker
 * collection on both success and failure, fileid conventions -- and most of
 * its comments record a bug that motivated the current shape. Where this file
 * differs from it, the difference is forced by the runtime and is commented as
 * such; everywhere else the Python source is the specification and the
 * tiebreaker.
 */
import {
  formatMarkers,
  resolveMarker,
  sortMarkers,
  type Marker,
  type RawMarker,
} from './Marker.js';
import { deserialize } from './ast/generated/deserialize.js';
import { GlobalScope, RootSymbolScope } from './ast/generated/index.js';
import { initPssParser, type ParserSessionHandle, type InitOptions } from './wasm/loader.js';

// `using` lowers to a Symbol.dispose lookup, and the well-known symbol is
// absent on runtimes predating the proposal's implementation (Node < 18.18).
// Defining it here rather than depending on a polyfill package keeps the
// published package dependency-free, which is worth a line of setup.
//
// This must run before the class below is *defined*, not merely before it is
// instantiated: a computed method key is evaluated when the class definition
// is evaluated, so a polyfill at the foot of the file would install the symbol
// after `[Symbol.dispose]` had already been keyed on `undefined`.
if (typeof (Symbol as { dispose?: symbol }).dispose === 'undefined') {
  Object.defineProperty(Symbol, 'dispose', {
    value: Symbol.for('Symbol.dispose'),
    configurable: false,
    writable: false,
  });
}

export interface ParserOptions extends InitOptions {
  /** Collect doc comments and attach them to the declarations they document,
   *  reachable as `getDocstring()` on any `ScopeChild`. */
  collectDocstrings?: boolean;
  /** Capture *every* comment, on statements as well as declarations, as
   *  `ScopeChild.comments`. Implies `collectDocstrings`. */
  collectComments?: boolean;
}

/** An in-memory source file. */
export interface SourceFile {
  /** The name reported in markers and in `fileMap()`. Need not be a real
   *  path -- an unsaved editor buffer has no path, and that is the case this
   *  API is shaped for. */
  name: string;
  content: string;
}

/**
 * Thrown when a parse or link produces any error-severity marker.
 *
 * `markers` carries the diagnostics collected *before* the throw, which is the
 * whole reason this class exists rather than a plain `Error`. Both Python
 * parse paths call `_collectMarkers` and then raise (`parser.py:106-109`), so
 * a caught exception is self-describing; `parser.markers` holds the same list.
 */
export class ParseException extends Error {
  readonly markers: readonly Marker[];

  constructor(message: string, markers: readonly Marker[] = []) {
    super(message);
    this.name = 'ParseException';
    this.markers = markers;
  }
}

/**
 * Instantiate the WASM core if it is not loaded yet, and build a `Parser`.
 *
 * Async because WASM instantiation is; there is no synchronous construction
 * path and there will not be one. Use `initPssParser()` to move the cost
 * somewhere you control.
 */
export async function createParser(opts: ParserOptions = {}): Promise<Parser> {
  const module = await initPssParser(opts);
  return new Parser(new module.ParserSession(), opts);
}

/**
 * Warns about a `Parser` that reached garbage collection without `dispose()`.
 *
 * A safety net, never a substitute (`ts-api-design.md` §9 Q1). Finalizers are
 * not guaranteed to run at all -- not on process exit, not under memory
 * pressure, not ever -- so this cannot free anything on a schedule a program
 * may depend on. What it can do is turn a silent leak into a message during
 * development, which is the only thing it is here for.
 *
 * It holds the *name*, never the `Parser`: a registry entry that referenced
 * its own target would keep the target alive and guarantee the callback never
 * fires.
 */
const leakRegistry =
  typeof FinalizationRegistry === 'undefined'
    ? null
    : new FinalizationRegistry<string>((label) => {
        console.warn(
          `pssparser: ${label} was garbage collected without dispose(). ` +
            'The WASM-side parser was leaked. Call dispose(), or declare the ' +
            'parser with `using`.'
        );
      });

let parserSeq = 0;

/**
 * Profiling data from the last parse.
 *
 * Deliberately opaque (`ts-api-design.md` §9 Q4). What the ANTLR profiler
 * reports is a diagnostic aid whose shape is not obviously stable, and
 * publishing a precise type would turn "we now also count X" into a breaking
 * API change. The keys emitted today are documented in `docs/typescript_api.rst`.
 */
export type ProfileInfo = Readonly<Record<string, unknown>>;

export class Parser {
  #session: ParserSessionHandle | null;
  #label: string;

  #collectDocstrings: boolean;
  #collectComments: boolean;

  /**
   * fileid -> source name for user files.
   *
   * Written as sources are parsed and read when markers are resolved. This is
   * `_filenames` (`parser.py:32`); `#fileMap` below is the snapshot `link()`
   * takes of it.
   */
  #filenames = new Map<number, string>();
  #fileMap = new Map<number, string>();

  /**
   * Markers accumulated across every parse and link on this Parser.
   *
   * Accumulation is Python's behaviour and is preserved deliberately, but see
   * `clearMarkers()`: Python's consumers are one-shot CLIs, and a long-lived
   * Parser re-parsing on every keystroke grows this without bound.
   */
  #markers: Marker[] = [];

  /** The linked root, materialised by `link()`. */
  #root: RootSymbolScope | null = null;

  /** @internal Use `createParser()`. */
  constructor(session: ParserSessionHandle, opts: ParserOptions = {}) {
    this.#session = session;
    this.#label = `Parser#${++parserSeq}`;

    this.#collectDocstrings = opts.collectDocstrings ?? false;
    this.#collectComments = opts.collectComments ?? false;
    session.setCollectDocstrings(this.#collectDocstrings);
    session.setCollectComments(this.#collectComments);

    leakRegistry?.register(this, this.#label, this);
  }

  /**
   * Stop reporting further errors for a file after `n`. `0` (the default) is
   * unlimited.
   *
   * A library caller wants every diagnostic; the cap is a terminal-output
   * affordance a CLI opts into. Only error-severity markers count against it,
   * and a file that reaches it gets one extra PSS029 marker announcing the
   * cutoff.
   */
  setMaxErrors(n: number): void {
    this.#require().setMaxErrors(n);
  }

  /**
   * Parse in-memory sources.
   *
   * Throws `ParseException` if any source produces an error-severity marker,
   * after collecting the markers -- the exception carries them and `markers`
   * is updated either way.
   *
   * Stops at the first source that fails, as Python does (`parser.py:138-141`).
   * That is not just imitation: one marker collector serves the whole call, so
   * after a failure `hasErrors()` stays true and later sources could not be
   * reported on independently anyway.
   *
   * Sources that parse cleanly join this Parser's environment and are visible
   * to later `parseSources()` calls; a source that fails does not.
   */
  parseSources(files: SourceFile[]): boolean {
    const session = this.#require();
    session.beginParse();

    for (const file of files) {
      // Record the name *before* parsing. The fileid is assigned by the
      // session as the unit is created, and markers emitted during that parse
      // already carry it -- resolving them needs the name to be in the map
      // already.
      const fileid = session.nextFileid();
      this.#filenames.set(fileid, file.name);

      if (session.parseSource(file.content)) {
        // Collect before reporting. A caught ParseException carries the
        // markers (parser.py:106-109), and `markers` is populated too.
        this.#absorbMarkers(session);
        throw new ParseException(formatMarkers(this.#markers), this.#markers);
      }
    }

    this.#absorbMarkers(session);
    return true;
  }

  /**
   * Markers from every parse and link performed on this Parser, sorted by
   * (file, line, col).
   *
   * A copy: the array a consumer holds must not change under it when the next
   * parse runs.
   */
  get markers(): readonly Marker[] {
    return this.#markers.slice();
  }

  /**
   * Drop the accumulated markers.
   *
   * Not in the Python API, and added for the case Python does not have
   * (`ts-api-design.md` §9 Q5). Python accumulates across calls on one Parser
   * and its consumers are one-shot CLIs, so nothing ever noticed. A language
   * server re-parsing the same buffer on every keystroke would notice: the
   * list grows without bound and every entry is a duplicate of the last pass.
   *
   * Accumulation stays the default, because changing it would change what
   * `link()` reports -- link-time markers are appended to parse-time ones on
   * purpose (`parser.py:180-184`).
   */
  clearMarkers(): void {
    this.#markers = [];
  }

  /** Whether doc comments are collected. Set at construction. */
  get collectDocstrings(): boolean {
    return this.#collectDocstrings;
  }

  /** Whether every comment is collected. Set at construction. */
  get collectComments(): boolean {
    return this.#collectComments;
  }

  /**
   * fileid -> source name for the user-supplied files.
   *
   * Populated by `link()` and valid afterwards -- including after a *failed*
   * link, which is the point (`parser.py:186-217`). Built-in and
   * standard-library units are excluded: they carry fileid -1 and 0 and have
   * no user path.
   */
  fileMap(): ReadonlyMap<number, string> {
    return new Map(this.#fileMap);
  }

  /** Enable ANTLR profiling for subsequent parse operations. */
  enableProfiling(enable = true): void {
    this.#require().setEnableProfiling(enable);
  }

  /**
   * Profiling data from the last parse, or null when there is none.
   *
   * Null covers three cases and does not distinguish them: profiling was off,
   * nothing has been parsed yet, or `link()` has run. That last one is a
   * deliberate divergence from Python, which keeps a `_last_builder` reference
   * and would report the profile of a builder it released at link time
   * (`parser.py:161-164`); the builder is gone here, so the answer is honestly
   * nothing rather than stale.
   */
  getProfileInfo(): ProfileInfo | null {
    const json = this.#require().profileInfoJson();
    // '' rather than 'null' from the binding, so the empty case costs no
    // JSON parse.
    return json === '' ? null : (JSON.parse(json) as ProfileInfo);
  }

  /**
   * Link every unit parsed so far and return the linked root.
   *
   * Throws `ParseException` if the link produces any error-severity marker --
   * after recording the root, which is the part that is easy to get backwards.
   *
   * Four behaviours here are load-bearing, and each is a bug `parser.py`
   * records in a comment:
   *
   *  1. Collect markers *unconditionally*, not only on failure. Collecting
   *     inside the error branch dropped every warning from a successful link,
   *     and the CLI printed "0 errors in 0 files" (`parser.py:174-179`).
   *  2. Re-sort the concatenation. Parse-time and link-time markers are each
   *     internally sorted; two sorted lists concatenated are not sorted
   *     (`parser.py:182-184`).
   *  3. Record the root *before* reporting failure. Ownership of the units has
   *     already moved into the root by then, so a failed link must still leave
   *     `root`, `userUnits()` and `fileMap()` usable (`parser.py:186-217`).
   *  4. Drop the builder. It holds borrowed pointers to the units as its
   *     compile-time environment, and ownership has just moved away
   *     (`parser.py:224-231`). The binding's `link()` does this; a later
   *     `parseSources()` builds a fresh one and reloads the standard library,
   *     because the unit list restarted too.
   *
   * The returned tree is a materialised snapshot: plain JavaScript objects
   * that survive `dispose()` and do not change under a later parse.
   */
  link(): RootSymbolScope {
    const session = this.#require();

    const failed = session.link();

    // (1) and (2): collect and re-sort whether or not the link failed.
    // #absorbMarkers does both, and resolves fileids against the live map --
    // which is why it must run before the snapshot-and-clear below, or every
    // link marker would resolve against a map that no longer holds the ids.
    this.#absorbMarkers(session);

    // (3): record the result before reporting failure.
    this.#root = this.#materialiseRoot(session);

    // Snapshot fileid -> name, then clear the live map. The clear is not
    // tidiness: the next parse restarts the unit list, so fileid 1 will name a
    // different file. #names() consults the live map first for exactly that
    // reason, and the snapshot keeps older markers resolvable
    // (`parser.py:_pathOf`).
    this.#fileMap = new Map(this.#filenames);
    this.#filenames.clear();

    if (failed) {
      throw new ParseException(formatMarkers(this.#markers), this.#markers);
    }
    return this.#root;
  }

  /** The linked root; null before `link()`. */
  get root(): RootSymbolScope | null {
    return this.#root;
  }

  /**
   * The `GlobalScope` of each user-supplied file, in parse order.
   *
   * Read from the linked root, which owns the units after `link()`. Built-in
   * and standard-library units are filtered out by fileid, which is what
   * `fileMap()` membership tests. Empty before `link()`.
   *
   * Does not require a live session: the tree is already materialised, so this
   * keeps working after `dispose()`.
   */
  userUnits(): GlobalScope[] {
    if (!this.#root) {
      return [];
    }
    return this.#root.units.filter((u) => this.#fileMap.has(u.fileid));
  }

  /**
   * Release the WASM-side parser.
   *
   * Required, not an optimisation: the WASM heap is not reachable by the
   * JavaScript garbage collector, so nothing else will free it. Idempotent --
   * calling it twice is not an error, because a `finally` that runs after an
   * explicit dispose is ordinary code. Every other method throws afterwards.
   *
   * AST objects already materialised are plain JavaScript and survive this.
   * That is deliberate: a language server can hold the last good AST while
   * disposing the parser that produced it.
   */
  dispose(): void {
    if (this.#session) {
      this.#session.delete();
      this.#session = null;
      leakRegistry?.unregister(this);
    }
  }

  /** Explicit resource management: `using parser = await createParser()`. */
  [Symbol.dispose](): void {
    this.dispose();
  }

  #require(): ParserSessionHandle {
    if (!this.#session) {
      throw new Error(`pssparser: ${this.#label} has been disposed`);
    }
    return this.#session;
  }

  /**
   * Pull the session's markers into `#markers`, resolved and re-sorted.
   *
   * Re-sorting the whole list rather than the new markers is not redundant.
   * Each collection is internally sorted, but concatenating two sorted lists
   * does not produce a sorted list -- the bug `parser.py:182-184` records.
   */
  #absorbMarkers(session: ParserSessionHandle): void {
    const raw = JSON.parse(session.markersJson()) as RawMarker[];
    const names = this.#names();
    // Markers from this call replace nothing; they are appended to whatever
    // earlier calls left, because `markers` spans the Parser's lifetime.
    for (const r of raw) {
      this.#markers.push(resolveMarker(r, names));
    }
    sortMarkers(this.#markers);
  }

  /**
   * fileid -> name for resolving a marker, across the `link()` boundary.
   *
   * `link()` snapshots `#filenames` into `#fileMap` and clears it, and the
   * next parse then reuses the same fileids for different files. So the live
   * map must win where the two disagree, and the snapshot must still be
   * consulted so a marker from before the link does not resolve to
   * `<unknown>`. This is `_pathOf` (`parser.py:276-287`), which consults both
   * in the same order and for the same reason.
   */
  #names(): ReadonlyMap<number, string> {
    if (this.#fileMap.size === 0) {
      return this.#filenames;
    }
    return new Map([...this.#fileMap, ...this.#filenames]);
  }

  /**
   * Pull the linked root across the boundary as plain JavaScript.
   *
   * One buffer carries the whole linked program: `RootSymbolScope` owns the
   * units (`list<UP<GlobalScope>>`) and the serialiser descends through owning
   * pointers. That is what makes a cross-unit reference resolvable after
   * linking -- before it, each unit was serialised alone and a pointer into
   * another unit had no id to write (`ts-api-design.md` §4).
   *
   * The bytes must be copied before anything else touches the WASM heap: what
   * the binding returns is a view over it, and `ALLOW_MEMORY_GROWTH` can
   * detach it outright.
   */
  #materialiseRoot(session: ParserSessionHandle): RootSymbolScope {
    const root = deserialize(new Uint8Array(session.serializeRoot()));
    if (!(root instanceof RootSymbolScope)) {
      // Not defensive padding: this would mean the linker returned something
      // other than a RootSymbolScope, or the wire format and the reader
      // disagree about the class tag. Either is a core or generator bug, and
      // both are far cheaper to see here than three field accesses later.
      throw new Error(
        `pssparser: link() produced ${root === null ? 'nothing' : root.constructor.name}, ` +
          'not a RootSymbolScope. The WASM core and the generated reader disagree.'
      );
    }
    return root;
  }
}

